app/src/main/java/ch/epfl/sdp/appart/AdActivity.java

```diff
@@ -192,18 +211,125 @@ public boolean onCreateOptionsMenu(Menu menu) {
+                Toast.makeText(this, e.getMessage(),
+                        Toast.LENGTH_SHORT).show();
+                return null;
             });
+            addRes.thenAccept(res ->
+                    Toast.makeText(this, R.string.favSuccess_Ad,
+                            Toast.LENGTH_SHORT).show()
+            );
             return true;
         }
         return super.onOptionsItemSelected(item);
     }
 
+    /**
+     * Check if user is logged in and add new favorite ad to user
+     */
+    private CompletableFuture<Void> addNewFavorite() {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        // check that the user is logged in
+        User user = login.getCurrentUser();
+        if (user == null) {
+            result.completeExceptionally(
+                    new UnsupportedOperationException(getString(R.string.userNotLoggedIn_Ad))
+            );
+            return result;
+        }
+        CompletableFuture<User> userRes = database.getUser(user.getUserId());
+        userRes.exceptionally(e -> {
+            result.completeExceptionally(
+                    new DatabaseServiceException(getString(R.string.favFail_Ad)));
+            return null;
+        });
+        userRes.thenAccept(u -> saveFavorite(result, u));
+        return result;
+    }
 
+    /**
+     * Save new ad id in database and if successful save ad locally
+     */
+    private void saveFavorite(CompletableFuture<Void> result, User user) {
+        user.addFavorite(adId);
+        CompletableFuture<Boolean> updateRes = database.updateUser(user);
+        updateRes.exceptionally(e -> {
+            result.completeExceptionally(
+                    new DatabaseServiceException(
+                            getString(R.string.favFail_Ad)));
+            return null;
+        });
+        // if update successful, save ad locally
+        updateRes.thenAccept(res -> {
+            List<Bitmap> images = getImages();
+            List<Bitmap> panoramas = new ArrayList<>();
+            List<CompletableFuture<Bitmap>> panoramasRes =
+                    getPanoramasFutures();
+            User currentUser = login.getCurrentUser();
+            CompletableFuture<Bitmap> pfpRes = getProfilePic(currentUser);
+            pfpRes.exceptionally(e -> {
+                Toast.makeText(this, R.string.localFavFail_Ad
+                        , Toast.LENGTH_SHORT).show();
+                return null;
+            });
+            pfpRes.thenAccept(pfp -> {
+                CompletableFuture<Void> allOf = CompletableFuture.allOf(panoramasRes.toArray(
+                        new CompletableFuture[panoramasReferences.size()]));
+                allOf.exceptionally(e -> {
+                    Toast.makeText(this,
+                            R.string.localFavFail_Ad,
+                            Toast.LENGTH_SHORT).show();
+                    return null;
+                });
+                allOf.thenAccept(ignoredRes -> {
```

Comment:
Just as a remark (it's fine) I think you can also combine futures, this would allow reducing nested callbacks.